### ISSUE TITLE
Proposal: loader instantiation fix/improvement

### ISSUE THREAD

**Issue by sijakret** [AssemblyScript/assemblyscript#780]
Hi Everyone,

When instantiating any real-world module (>4KB) using the loader the following error comes up (at the very least in chrome):

> RangeError: WebAssembly.Instance is disallowed on the main thread, if the buffer size is larger than 4KB. Use WebAssembly.instantiate.

This is the code segment that causes the issue (lib/loader/index.js):
```js
/** Instantiates an AssemblyScript module using the specified imports. */
function instantiate(module, imports) {
  return postInstantiate(
    preInstantiate(imports || (imports = {})),
    new WebAssembly.Instance(module, imports)
  );
}
```
If i monkey patch it like below it  becomes usable (albeit async)
```js
/** Instantiates an AssemblyScript module using the specified imports. */
async function instantiate(module, imports) {
  return postInstantiate(
    preInstantiate(imports || (imports = {})),
    await WebAssembly.instantiate(module, imports)
  );
}
```

I would assume that this issue will be faced by others..

I see two easy ways to fix it:
 1. my patch above
 2. export `postInstantiate(..)` and `preInstantiate(..)` hooks so users can build costum instantiation flows.

What do you guys think?

**Comment by sijakret** [comment]
Has no one else encountered this?

**Comment by dcodeIO** [comment]
Iirc the expectation there was that one would use `instantiateStreaming` in such a context, but that might not cover every use case. If so, I suggest adding a separate `instantiateAsync` using your patch.

**Comment by sijakret** [comment]
I encountered the issue in the context I mentioned above. The reason I suggested option 2 is because it is extremely uninvasive = low risk. Since the instantiate method itself is so high level and simple in flow, just exporting pre.. and post.., would allow for a custom flow there and might be the easiest fix.

**Comment by dcodeIO** [comment]
Sure, I don't see any issues with exporting these.

**Comment by MaxGraey** [comment]
Should be fixed in #916
